src/analyzer.rs

```diff
@@ -1,6 +1,86 @@
 use crate::parser::Expr;
-use std::future::poll_fn;
+use crate::parser::ExprKind::{Binary, Bool, Float, Ident, InlineFunction, Int, Set, String};
+use miette::{miette, LabeledSpan, Severity};
 
-pub fn analyzer(exprs: Vec<Expr>) {
-    println!("{exprs:?}")
+pub fn analyze(source: &std::string::String, expressions: Vec<Expr>) {
+    let mut declared: Vec<(std::string::String, std::ops::Range<usize>)> = Vec::new();
+    let mut used: Vec<(std::string::String, std::ops::Range<usize>)> = Vec::new();
+
+    for expression in expressions {
+        let mut current_expr = expression;
+        let mut other: Option<Box<Expr>> = None;
```

Comment:
i don't really get why are you doing it this way, why not use a recursive function?

Reply:
because most of the times, using loops is faster than recursion

Reply:
Yes, no need for a recursive function here.